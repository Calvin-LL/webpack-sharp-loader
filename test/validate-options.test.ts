import { Sharp } from "sharp";
import webpack from "webpack";

import WSLWebpackTestCompiler from "./helpers/WSLWebpackTestCompiler";

describe.each([4, 5] as const)("v%d validate options", (webpackVersion) => {
  const tests = {
    processFunction: {
      success: [(sharp: Sharp) => sharp.flip()],
      failure: [true, undefined, () => undefined],
    },
    toBuffer: {
      success: [true, false],
      failure: [1],
    },
    fileLoaderOptions: {
      success: [{}],
      failure: [1],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ): void {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = new WSLWebpackTestCompiler({ webpackVersion });

      let stats: webpack.Stats | undefined;

      try {
        stats = (
          await compiler.compile({
            loaderOptions: {
              processFunction: (sharp: Sharp) => sharp.flip(),
              [key]: value,
            },
            throwOnError: false,
          })
        ).stats;
      } finally {
        if (type === "success") {
          expect(stats!.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const errors = stats!.compilation.errors;

          expect(errors).toHaveLength(1);
          expect(errors[0].error.message).toMatchSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
