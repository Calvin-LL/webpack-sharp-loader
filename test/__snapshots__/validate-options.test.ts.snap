// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v4 validate options should throw an error on the "fileLoaderOptions" option with 1 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from /Users/Calvin/GitHub/webpack-sharp-loader/test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoaderOptions should be an object:
   object { … }
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v4 validate options should throw an error on the "processFunction" option with true value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from /Users/Calvin/GitHub/webpack-sharp-loader/test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.processFunction should be an instance of function.
   -> A function that returns a sharp object.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v4 validate options should throw an error on the "processFunction" option with undefined value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from /Users/Calvin/GitHub/webpack-sharp-loader/test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options misses the property 'processFunction'. Should be:
   function
   -> A function that returns a sharp object.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v4 validate options should throw an error on the "processFunction" option with undefined value 2`] = `
Array [
  [ModuleBuildError: Module build failed (from /Users/Calvin/GitHub/webpack-sharp-loader/test-dist/cjs.js):
Error: processFunction must return a sharp object
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:49:11)],
]
`;

exports[`v4 validate options should throw an error on the "toBuffer" option with 1 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from /Users/Calvin/GitHub/webpack-sharp-loader/test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.toBuffer should be a boolean.
   -> Whether to output as buffer. Default to true.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v5 validate options should throw an error on the "fileLoaderOptions" option with 1 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoaderOptions should be an object:
   object { … }
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v5 validate options should throw an error on the "processFunction" option with true value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.processFunction should be an instance of function.
   -> A function that returns a sharp object.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v5 validate options should throw an error on the "processFunction" option with undefined value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options misses the property 'processFunction'. Should be:
   function
   -> A function that returns a sharp object.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;

exports[`v5 validate options should throw an error on the "processFunction" option with undefined value 2`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
Error: processFunction must return a sharp object
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:49:11)],
]
`;

exports[`v5 validate options should throw an error on the "toBuffer" option with 1 value 1`] = `
Array [
  [ModuleBuildError: Module build failed (from ../../test-dist/cjs.js):
ValidationError: Invalid options object. Sharp Loader has been initialized using an options object that does not match the API schema.
 - options.toBuffer should be a boolean.
   -> Whether to output as buffer. Default to true.
    at Object.validate (/Users/Calvin/GitHub/webpack-sharp-loader/node_modules/schema-utils/dist/validate.js:104:11)
    at Object.default_1 (/Users/Calvin/GitHub/webpack-sharp-loader/src/index.ts:36:3)],
]
`;
